man/man3/moment.3
.TH MOMENT 3NEMO "18 January 2025"
.SH NAME
ini_moment, slow_moment, accum_moment, decr_moment, 
reset_moment, show_moment, n_moment, sum_moment, sratio_moment,
mean_moment, sigma_moment, skewness_moment, kurtosis_moment, mad_moment, mard_moment, robust_moment,
min_moment, max_moment \- various (moving) moment and minmax routines
.SH SYNOPSIS
.nf
.B
#include <moment.h>
.PP
.B void ini_moment(m, mom, ndat)
.B void slow_moment(m, slow)
.B void accum_moment(m, x, w)
.B void decr_moment(m, x, w)
.B void reset_moment(m)
.PP
.B real show_moment(m, mom)
.B int n_moment(m)
.B real sum_moment(m)
.B real sratio_moment(m)
.B real mean_moment(m)
.B real median_moment(m)
.B real sigma_moment(m)
.B real skewness_moment(m)
.B real kurtosis_moment(m)
.B real h3_moment(m)
.B real h4_moment(m)
.B real mad_moment(m)
.B real mard_moment(m)
.PP
.B real min_moment(m);
.B real max_moment(m);
.PP
.B void compute_robust_moment(m);
.B int n_robust_moment(m);
.B real min_robust_moment(m);
.B real max_robust_moment(m);
.B real mean_robust_moment(m);
.B real median_robust_moment(m);
.B real sigma_robust_moment(m);
.PP
.B Moment *m;
.B int mom, ndat;
.B real x, w;
.fi
.SH DESCRIPTION
\fImoment\fP is a set of functions to compute the moments of 
a set of real values, but also keep track of the datamin
and datamax. The routines are written in C but in 
an object-oriented fashion as to try and keep the use from the internal
data-structures. The \fBMoment\fP can (and should) be treated as an
opaque datastructure. All information should be obtained through the
interface routines described below.
.PP
If only a datamin/max is needed, setting \fBmom<0\fP can be used to prevent
the more expensive moment calculations.  See also \fBslow_moment\fP.
.PP
Moving or running averages (or moments) can be done by supplying \fBndat>0\fP
to ini_moment. It will keep a memory of the last \fBndat\fP data values
and the moments now become running moments. This is also the only way
the slow_moment (more accurate for large values) will work.
.PP
Note that the \fImedian_moment\fP can only be used in \fBx\fP (the weights are
ignored) and moving moment where \fBndat>0\fP.
.PP
\fBslow_moment\fP can be set with \fBslow=TRUE\fP to force re-computing higher order
moments based on the value of the mean. This only will work when \fBndat>0\fP.
.PP
\fBmean_moment\fP returns the mean value, where \fBsigma_moment\fP returns
the square root of the variance
.PP
\fBsratio_moment\fP returns a measure of how much signal there is compared to
the noise, by compting (SumP+SumN)/(SumP-SumN). For pure noise  this should be near 0,
and for pure signal, where SumN=0, near 1.
.PP
\fBrobust_moment's\fP can also be computed. These are based on an algorithm to
remove the outliers. It first computes the distance between the 1st and 3rd quartile,
then removes all points further than 1.5 times that distance from the 1st and 
3rd quartile. After calling \fBcompute_robust_moment\fP, the mean, median and
sigma can be retrieved as those computations are done in a temporary Moment
structure. Caveat: for large datasets and/or moving moments, 
robust_moment's can become expensive because of the need for sorting to
get the quartiles.
.PP
\fBmad_moment\fP computes the Median Absolute Deviation (MAD), arguably a better
measure for the Standard Deviation. As with the robust moments, it needs to
keep a copy of the data available. MAD is formally RMS/1.4826.  Related is
\fBmard_moment\fP, the Mean Absolute Relative Difference (MARD).
.SH MOMENTS
A note on the h3 and h4 moments, somewhat peculiar to astronomy. See
S2.4 in van der Marel & Franx (1993) 
.SH EXAMPLES
The following code computes a weighted mean and dispersion of a set of points:
.nf

    real x[100], w[100];
    int  i,n=100;
    Moment m;
    ...
    ini_moment(&m,2,0);      /* up to 2nd order moment, and using no circular buffer */
    for (i=0; i<n; i)
	accum_moment(&m,x[i],w[i]);
    printf("Mean: %g   Dispersion: %g\\n",
	mean_moment(&m), sigma_moment(&m));
    compute_robust_moment(&m);
    printf("Robust Mean: %g   Dispersion: %g\\n",
	mean_robust_moment(&m), sigma_robust_moment(&m));

.fi
.SH MOMENT STRUCTURE
A simple data structure (referred to as \fImoment\fP in the above
SYNOPSIS) is used to communicate between different routines:
.nf
.ta +0.3i +1.5i

typedef struct { 
	int mom;
	int n;
	real *sum;
	real datamin, datamax;

	int ndat;
	int idat;
	real *dat;
	real *sum;
} Moment;

.fi
from the standard include file \fImoment.h\fP.
.SH BUGS
When \fIdecr_moment\fP is used, the data min/max is not correct. 
Only with \fBndat>0\fP for moving moments can it be recomputed
correctly.
.SH SEE ALSO
grid(3NEMO)
.nf
http://apophenia.info
.fi
.SH AUTHOR
Peter Teuben
.SH FILES
.nf
.ta +1.5i
~/src/kernel/misc	moment.c
.fi
.SH UPDATE HISTORY
.nf
.ta +1.25i +4.5i
30-oct-93	Created   	PJT
8-nov-93	fixed init bug 	PJT
13-jun-95	added decr_moment	PJT
2-feb-05	added moving moments	PJT
2-mar-11	added h3,h4	PJT
24-apr-13	documented robust statistics	PJT
16-jan-14	added MAD	PJT
11-jun-14	clarified MAD and MARD (the old MAD was really MARD)	PJT
12-jul-20	added min/max for robust moment		PJT
14-nov-21	added sratio	PJT
18-jan-25	added slow option	PJT
.fi


src/kernel/misc/moment.c
/*
 * MOMENTS: accumulate and compute moments - almost OO !
 *
 *  30-oct-93   created         peter teuben
 *   9-nov	reset at ini 
 *   4-mar-94   ansi
 *  12-apr-95   return 0 for mean, sigma etc. when no points accumulated
 *  13-jun-95   added decr_moment, but min/max become invalid
 *  16-sep-95   fixed more bugs if no data accumulated
 *   9-dec-99   fix min/max when data deleted
 *   2-feb-05   added moving moments
 *  12-dec-07   added rms_moment; only valid for equal weights
 *  18-mar-11   robust mean, 
 *   9-oct-12   free_moment()  
 *  23-apr-13   add compute_robust_moment() 
 *   5-jun-13   add robust_range()
 *  15-jan-14   add MAD (mean absolute deviation)
 *  11-jun-14   MAD is really median absolute deviation?  MAD0 for now the old mean
 *              is that MARD (mean absolute relative difference)
 *  12-jul-20   add min/max for robust
 *  10-oct-20   median improvement via inline sort
 *  14-nov-21   add sratio
 *     feb-25   implement slow mode
 *
 * @todo    iterative robust by using a mask
 *          ? robust factor, now hardcoded at 1.5
 */


#include <stdinc.h>
#include <moment.h>

#define sum0 m->sum[0]
#define sum1 m->sum[1]
#define sum2 m->sum[2]
#define sum3 m->sum[3]
#define sum4 m->sum[4]

/* median.c */
extern real smedian(int,real*);
extern real smedian_q1(int,real*);
extern real smedian_q3(int,real*);
extern real pmedian(int,real*);
extern real pmedian_q1(int,real*);
extern real pmedian_q3(int,real*);

void ini_moment(Moment *m, int mom, int ndat)
{
    int i;

    m->n = 0;
    m->mom = mom;
    if (mom < 0)
      m->sum = NULL;
    else {
      m->sum = (real *) allocate((mom+1) * sizeof(real));
      for (i=0; i<=mom; i++) m->sum[i] = 0.0;
    }
    m->ndat = ndat;
    dprintf(1,"ini_moment: mom=%d ndat=%d\n", mom, ndat);

    if (ndat > 0) {     /* moving moments */
      m->idat = -1;
      m->dat = (real *) allocate(ndat*sizeof(real));
      m->wgt = (real *) allocate(ndat*sizeof(real));
    }

    m->sumn = m->sump = 0.0;
    m->slow = FALSE;    
}

void free_moment(Moment *m)
{  
   if (m->ndat) {
     if (m->sum) free(m->sum);
     free(m->dat);
     free(m->wgt);
   }
}

void slow_moment(Moment *m, bool slow)
{
  if (m->ndat > 0) {
    m->slow = slow;
  } else {
    if (slow)
      error("slow mode cannot be set with ndat=0");
    m->slow = FALSE;    
  }
}

void accum_moment(Moment *m, real x, real w)
{
    real xx, sum = w;
    int i;

    if (m->n == 0) {
	m->datamin = m->datamax = x;
    } else {
    	m->datamin = MIN(x, m->datamin);
    	m->datamax = MAX(x, m->datamax);
    }
    m->n++;
    if (m->mom < 0) return;
    for (i=0; i <= m->mom; i++) {
        m->sum[i] += sum;
        sum *= x;
    }
    if (x<0) m->sumn += x;   /* formally should use diff from mean */
    if (x>0) m->sump += x;
    if (m->ndat > 0) {                   /* if moving moments .... */
      if (m->idat < 0)                        /* first time around */
	m->idat=0;  
      else if (m->n <= m->ndat)             /* buffer not full yet */
	m->idat++;
      else {                            /* buffer full, remove old */
	m->idat++;
	if (m->idat == m->ndat) m->idat = 0;      /* new point loc */

	xx = m->dat[ m->idat ];            /* remove the old point */
	sum = m->wgt[ m->idat ];
	dprintf(2,"del: %g\n",xx);
	for (i=0; i<= m->mom; i++) {
	  m->sum[i] -= sum;
	  sum *= xx;
	}
	m->n--;
      }
      dprintf(2,"add: %d %g\n",m->idat,x);

      m->dat[m->idat] = x;
      m->wgt[m->idat] = w;

    }
}


void decr_moment(Moment *m, real x, real w)
{
    real sum = w;
    int i;

    if (m->ndat > 0) 
      error("decr_moment: cannot be used in moving moments mode");

    if (m->n == 0) {
	warning("Cannot decrement a moment with no data accumulated");
    } else {
	/* note we cannot correct min/max here ....  client must update */
    }
    m->n--;
    if (m->mom < 0) return;
    for (i=0; i <= m->mom; i++) {
        m->sum[i] -= sum;
        sum *= x;
    }
    if (x<0) m->sumn -= x;
    if (x>0) m->sump -= x;
}

void reset_moment(Moment *m)
{
    int i;
    
    m->n = 0;
    m->idat = -1;
    if (m->mom < 0) return;
    for (i=0; i <= m->mom; i++)
        m->sum[i] = 0.0;
    m->sumn = m->sump = 0;
}

real show_moment(Moment *m, int mom)
{
    if (m->mom < 0) error("Cannot show moment for mom=%d",m->mom);
    if (mom >= 0 && mom <= m->mom)
        return m->sum[mom];
    else if (mom == -1)
        return mean_moment(m);
    else if (mom == -2)
        return sigma_moment(m);
    else if (mom == -3)
        return skewness_moment(m);
    else if (mom == -4)
        return kurtosis_moment(m);
    else 
        warning("show_moment: out-of-range moment %d %d",mom);
    return 0.0;
}

int n_moment(Moment *m)
{
    return m->n;
}

real sum_moment(Moment *m)
{
    return sum1;
}

real sratio_moment(Moment *m)
{
    return (m->sump + m->sumn)/(m->sump - m->sumn);
}

real mean_moment(Moment *m)
{
    if (m->mom < 1)
        error("mean_moment cannot be computed with mom=%d",m->mom);
    if (m->n == 0) return 0;
    return sum1/sum0;
}

/* 
 * @todo http://en.wikipedia.org/wiki/Chauvenet's_criterion
 *       http://en.wikipedia.org/wiki/Robust_statistics
 *       implement IDL's 'where' masking (see Moment->msk)
 *
 * @todo these temp variables should be stored inside the Moment structure
 */

static real  last_min_robust_moment    = -1;
static real  last_max_robust_moment    = -1;
static real  last_mean_robust_moment   = -1;
static real  last_sigma_robust_moment  = -1;
static real  last_median_robust_moment = -1;
static int   last_n_robust_moment      = -1;
static real  last_robust_range[2];


local int compar_real(const void *va, const void *vb)
{
  real *a = (real *) va;
  real *b = (real *) vb;
  return *a < *b ? -1 : *a > *b ? 1 : 0;
}

local real get_median(int n, real *x)
{
  dprintf(1,"get_median: n=%d\n",n);

  qsort(x,n,sizeof(real),compar_real);
  if (n % 2)
    return  x[(n-1)/2];
  else
    return 0.5 * (x[n/2] + x[n/2-1]);
}



void compute_robust_moment(Moment *m)
{
  int i,n;
  real m1,m2,m3,iqr,dlo,dhi;
  Moment tmp;
  real frob = 1.5;   /* hardcoded for now */

  if (m->ndat==0)
    error("mean_robust_moment cannot be computed with ndat=%d",m->ndat);
  n = MIN(m->n, m->ndat);
#if 0
  m2 = pmedian(n,m->dat);
  m1 = pmedian_q1(n,m->dat);
  m3 = pmedian_q3(n,m->dat);
#else
  get_median(n,m->dat);    // sort, but this will destroy the weights array
  m2 = smedian(n,m->dat);
  m1 = smedian_q1(n,m->dat);
  m3 = smedian_q3(n,m->dat);  
#endif  
  iqr = m3-m1;
  dlo = m1 - frob*iqr;   /* perhaps better if this 1.5 factor */
  dhi = m3 + frob*iqr;   /* should depend on the # datapoints */
  ini_moment(&tmp,2,n);
  for (i=0; i<n; i++) {
    if (m->dat[i]<dlo || m->dat[i]>dhi) continue;
    accum_moment(&tmp,m->dat[i],1.0);
  }
  dprintf(1,"compute_robust_moment: %d %g %g %g\n", n, m1, m2, m3);

  // speed up computation if recalled without changing data
  last_min_robust_moment    = min_moment(&tmp);
  last_max_robust_moment    = max_moment(&tmp);
  last_mean_robust_moment   = mean_moment(&tmp);
  last_sigma_robust_moment  = sigma_moment(&tmp);
  last_median_robust_moment = median_moment(&tmp);
  last_n_robust_moment      = n_moment(&tmp);
  last_robust_range[0]      = dlo;
  last_robust_range[1]      = dhi;
  free_moment(&tmp);
}

// signal = np.median(flux)
// noise  = 0.6052697 * np.median(np.abs(2.0 * flux[2:n-2] - flux[0:n-4] - flux[4:n]))
// return signal / noise
#if 0
void compute_der_snr(Moment *m)
{
  int i,n;
  real m1,m2,m3,iqr,dlo,dhi;
  Moment tmp;
  real fac = 0.6052697;
  real ne = 2;
  real *qdat;
  
  if (m->ndat==0)
    error("mean_robust_moment cannot be computed with ndat=%d",m->ndat);
  
  n = MIN(m->n, m->ndat);
  m2 = median(n,m->dat);

  qdat = (real *) allocate((n-ne) * sizeof(real));
  for (i=ne; i<n-ne; i++) {
    qdat[i] = 2*m->dat[i] - m->dat[i-ne] - m->dat[i+ne];
  }

  
  m1 = median_q1(n,m->dat);
  m3 = median_q3(n,m->dat);
  iqr = m3-m1;
  dlo = m1 - frob*iqr;   /* perhaps better if this 1.5 factor */
  dhi = m3 + frob*iqr;   /* should depend on the # datapoints */
  ini_moment(&tmp,2,n);
  for (i=0; i<n; i++) {
    if (m->dat[i]<dlo || m->dat[i]>dhi) continue;
    accum_moment(&tmp,m->dat[i],1.0);
  }
  last_mean_robust_moment   = mean_moment(&tmp);
  last_sigma_robust_moment  = sigma_moment(&tmp);
  last_median_robust_moment = median_moment(&tmp);
  last_n_robust_moment      = n_moment(&tmp);
  last_robust_range[0]      = dlo;
  last_robust_range[1]      = dhi;
  free_moment(&tmp); 
}

#endif

int n_robust_moment(Moment *m)
{
  return last_n_robust_moment;
}

real mean_robust_moment(Moment *m)
{
  return last_mean_robust_moment;
}

real min_robust_moment(Moment *m)
{
  return last_min_robust_moment;
}
real max_robust_moment(Moment *m)
{
  return last_max_robust_moment;
}

real median_robust_moment(Moment *m)
{
  return last_median_robust_moment;
}

real sigma_robust_moment(Moment *m)
{
  return last_sigma_robust_moment;
}

void robust_range(Moment *m, real *range)
{
  range[0] = last_robust_range[0];
  range[1] = last_robust_range[1];
}

real median_moment(Moment *m)
{
  int n;
  if (m->ndat==0)
    error("median_moment cannot be computed with ndat=%d",m->ndat);
  dprintf(1,"median_moment: n=%d ndat=%d\n",m->n, m->ndat);
  n = MIN(m->n, m->ndat);
  return smedian(n,m->dat);
}

real sigma_moment(Moment *m)
{
    real mean, tmp2;
    
    if (m->mom < 2)
        error("sigma_moment cannot be computed with mom=%d",m->mom);
    if (m->n == 0) return 0.0;
    mean=sum1/sum0;
    if (m->datamin == m->datamax) return 0.0;
    if (m->slow) {
      if (m->ndat==0)
	error("slow sigma cannot be computed with ndat=%d",m->ndat);
      tmp2 = 0.0;
      for (int i=0; i<m->ndat; i++) {
	tmp2 += sqr(m->dat[i]-mean);
      }
      tmp2 = tmp2/sum0;
    } else
      tmp2 = sum2/sum0 - mean*mean;
    if (tmp2 <= 0.0) return 0.0;
    return sqrt(tmp2);
}

// MARD = Mean Absolute Relative Difference

real mard_moment(Moment *m)
{
  real mean, x;
  int i, n;
  Moment tmp;

  if (m->ndat==0)
    error("mard_moment cannot be computed with ndat=%d",m->ndat);
  mean = sum1/sum0;
  n = MIN(m->n, m->ndat);
  ini_moment(&tmp,1,n);
  for (i=0; i<n; i++) {
    x = m->dat[i] - mean;
    if (x > 0)
      accum_moment(&tmp,x,1.0);
    else
      accum_moment(&tmp,-x,1.0);
  }
  mean = mean_moment(&tmp);
  free_moment(&tmp);
  return mean;
}

/*
 *  MAD = Median Absolute Deviation
 *  (sigma = 1.4826 * mad for a normal distribution)
 */

real mad_moment(Moment *m)
{
  real median, x;
  int i, n;
  Moment tmp;

  if (m->ndat==0)
    error("mad_moment cannot be computed with ndat=%d",m->ndat);
  median = median_moment(m);
  n = MIN(m->n, m->ndat);
  ini_moment(&tmp,1,n);
  for (i=0; i<n; i++) {
    x = m->dat[i] - median;
    if (x > 0)
      accum_moment(&tmp,x,1.0);
    else
      accum_moment(&tmp,-x,1.0);
  }
  median = median_moment(&tmp);
  free_moment(&tmp);
  return median;
}


/* TODO:  this needs to be weight independant 
 *        now it only works properly if all weighta are 1
 */

real rms_moment(Moment *m)
{
    real mean, tmp;
    if (m->mom < 2)
        error("rms_moment cannot be computed with mom=%d",m->mom);
    if (m->n < 2) return 0;
    mean=sum1/sum0;
    if (m->datamin == m->datamax) return 0.0;
    tmp = sum2 - mean*mean*sum0;
    if (tmp <= 0.0) return 0.0;
    tmp /= (m->n - 1);
    return sqrt(tmp);
}

real skewness_moment(Moment *m)
{
    real mean, sigma, tmp, tmp2, tmp3;

    if (m->mom < 3)
        error("skewness_moment cannot be computed with mom=%d",m->mom);
    if (m->n == 0) return 0.0;
    if (m->datamin == m->datamax) return 0.0;
    mean = sum1/sum0;
    if (m->slow) {
      tmp2 = tmp3 = 0.0;
      for (int i=0; i<m->ndat; i++) {
	tmp2 += sqr(m->dat[i]-mean);
	tmp3 += qbe(m->dat[i]-mean);
      }
      sigma = tmp2/sum0;
      if (sigma < 0.0) sigma = 0.0;
      sigma = sqrt(sigma);
      tmp = (tmp3/sum0 ) / (sigma*sigma*sigma);      
    } else {
      sigma = sum2/sum0 - mean*mean;
      if (sigma < 0.0) sigma = 0.0;
      sigma = sqrt(sigma);    
      tmp = ((sum3-3*sum2*mean)/sum0 + 2*mean*mean*mean) /
	(sigma*sigma*sigma);
    }
    return tmp;
}

real kurtosis_moment(Moment *m)
{
    real mean, sigma2, tmp, tmp2, tmp4;

    if (m->mom < 4)
    error("kurtosis_moment cannot be computed with mom=%d",m->mom);
    if (m->datamin == m->datamax) return 0.0;
    if (m->n == 0) return 0.0;
    mean = sum1/sum0;
    if (m->slow) {
      tmp2 = tmp4 = 0.0;
      for (int i=0; i<m->ndat; i++) {
	tmp2 += sqr(m->dat[i]-mean);
	tmp4 += sqr(sqr(m->dat[i]-mean));
      }
      sigma2 = tmp2/sum0;
      tmp = -3.0 + (tmp4/sum0 ) / (sigma2*sigma2);
    } else {
      sigma2 = sum2/sum0 - mean*mean;
      tmp = -3.0 +
	    ((sum4-4*sum3*mean+6*sum2*mean*mean)/sum0 - 3*mean*mean*mean*mean) /
            (sigma2*sigma2);
    }
    return tmp;
}

/* for h3 and h4, see S2.4 in van der Marel & Franx (1993) */
/* skew: zeta_1 = mu_3 / mu_2^(3/2)  = 0 for a gaussian    */
/* kurt: zeta_2 = mu_4 / mu_2^2      = 3 for a gaussian    */
/*       zeta_1 = 4 sqrt(3) h_3      approx */
/*       zeta_2 = 3 + 8 sqrt(6) h_4  approx */

real h3_moment(Moment *m)
{
  return skewness_moment(m) / (4*sqrt(3.0));
}

real h4_moment(Moment *m)
{
  return kurtosis_moment(m) / (8*sqrt(6.0));
}



real min_moment(Moment *m)
{
  int i, n;
  if (m->ndat > 0) {
    n = MIN(m->ndat, m->n);
    m->datamin = m->dat[0];
    for (i=1; i<n; i++)
      m->datamin = MIN(m->dat[i], m->datamin);
  }
  return m->datamin;
}                                        

real max_moment(Moment *m)
{
  int i, n;
  if (m->ndat > 0) {
    n = MAX(m->ndat, m->n);
    m->datamax = m->dat[0];
    for (i=1; i<n; i++)
      m->datamax = MAX(m->dat[i], m->datamax);
  }    
  return m->datamax;
}                                        

#ifdef TESTBED
#include <getparam.h>

string defv[] = {
    "in=???\n       Table with values in column 1",
    "moment=-1\n    Moment to compute (-4..-1 special, 0,1,2,...)",
    "minmax=f\n     Show datamin & max instead ? ",
    "median=f\n     Show median ?",
    "robust=f\n     Show robust mean etc.?",
    "maxsize=0\n    If > 0, size for moving moments instead\n",
    "VERSION=0.4\n  23-apr-2013 PJT",
    NULL,
};

string usage = "(Moving)Moments TESTBED";

void debug_moment(int d, Moment *m)
{
  int i, n;
  if (m->ndat == 0) return;

  n = MIN(m->ndat, m->n);
  dprintf(d,"moment data[%d]: ",n);
  for (i=0; i<n; i++)
    dprintf(d,"%g ",m->dat[i]);
  dprintf(d,"\n");
}

void nemo_main(void)
{
    char line[80];
    stream instr = stropen(getparam("in"),"r");
    int mom = getiparam("moment");
    int maxsize = getiparam("maxsize");
    real x;
    Moment m;
    bool Qminmax = getbparam("minmax");
    bool Qmedian = getbparam("median");
    bool Qrobust = getbparam("robust");

    ini_moment(&m,ABS(mom),maxsize);
    while (fgets(line,80,instr) != NULL) {
      x = atof(line);
      accum_moment(&m,x,1.0);
      if (maxsize > 0) {
	debug_moment(1,&m);
	printf("%d %g ",n_moment(&m),x);
	if (Qminmax)
	  printf("%g %g\n",min_moment(&m), max_moment(&m));
	else if (Qmedian)
	  printf("%g\n",median_moment(&m));
	else if (Qrobust) {
	  compute_robust_moment(&m);
          printf("%g\n",mean_robust_moment(&m));
	} else
	  printf("%g\n",show_moment(&m,mom));
      }
    }
    if (maxsize == 0) {
      printf("%d %g ",n_moment(&m),x);
      if (Qminmax)
        printf("%g %g\n",min_moment(&m), max_moment(&m));
      else if (Qmedian)
	printf("%g\n",median_moment(&m));
      else if (Qrobust)  {
	compute_robust_moment(&m);
        printf("%g\n",mean_robust_moment(&m));
      } else
        printf("%g\n",show_moment(&m,mom));
    }
}

#endif
